pylint/checkers/refactoring/refactoring_checker.py

```diff
@@ -684,18 +708,40 @@ def _check_consider_using_comprehension_constructor(self, node):
                 message_name = "consider-using-set-comprehension"
                 self.add_message(message_name, node=node)
 
+    def _check_consider_using_generator(self, node):
+        checked_call = ["any", "all"]
```

Comment:
I think @ngie-eign comment was relevant here https://github.com/PyCQA/pylint/pull/3309#discussion_r366091653
What do you think ?

Reply:
I am not sure, I have tested those.
The result is slower with the generator.

list:
```
python -m timeit "some_list = list(range(1000));b = list([x % 7 == 0 for x in some_list]);"
# output: 10000 loops, best of 3: 79.4 usec per loop

python -m timeit "some_list = list(range(1000));b = list(x % 7 == 0 for x in some_list);"
# output: 10000 loops, best of 3: 99.4 usec per loop
```

set:
```
python -m timeit "some_list = list(range(1000));b = set([x % 7 == 0 for x in some_list]);"
# output: 10000 loops, best of 3: 89.1 usec per loop

python -m timeit "some_list = list(range(1000));b = set(x % 7 == 0 for x in some_list);"
# output: 10000 loops, best of 3: 104 usec per loop
```

tuple:
```
python -m timeit "some_list = list(range(1000));b = tuple([x % 7 == 0 for x in some_list]);"
# output: 10000 loops, best of 3: 80.3 usec per loop

python -m timeit "some_list = list(range(1000));b = tuple(x % 7 == 0 for x in some_list);"
# output: 10000 loops, best of 3: 99.6 usec per loop
```

compared to `any` and `all`

any:
```
python -m timeit "some_list = list(range(1000));b = any([x % 7 == 0 for x in some_list]);"
# output: 10000 loops, best of 3: 74.2 usec per loop

python -m timeit "some_list = list(range(1000));b = any(x % 7 == 0 for x in some_list);"
# output: 100000 loops, best of 3: 17 usec per loop
```

all:
```
python -m timeit "some_list = list(range(1000));b = all([x % 7 == 0 for x in some_list]);"
# output: 10000 loops, best of 3: 84.6 usec per loop

python -m timeit "some_list = list(range(1000));b = all(x % 7 == 0 for x in some_list);"
# output: 100000 loops, best of 3: 16.5 usec per loop
```

When I checked the python docs, there `any` and `all` take `iterable` as the parameter. But `list`, `set`, and `tuple` takes `[iterable]`. And there isn't any other function that takes `iterable`

Reply:
Ok it makes senses. 
```suggestion
        # We only check 'any' and 'all' because for list, set, and tuple a generator performs worse
        # See https://github.com/PyCQA/pylint/pull/3309#discussion_r576683109
        checked_call = ["any", "all"]
```

Reply:
I agree with the timing concerns, but when it boils down to it, it's the unnecessary data copies that I'm worried about. With particularly large sequences or iterables, creating unnecessary copies really adds up.

Reply:
@ikraduya: out of curiosity, what version of python was your output above tested on?

Reply:
I think here we could create two separate messages. Actually for `all` or `any` there is nothing to 'consider', changing for a generator is a no-brainer, you can cut the execution tree and exit directly at the first element. Except in the worst possible case it's an easy performance win.  So we could have a stronger `use-generator-instead` message for this. Then for other container, maybe for really long lists or sets it would be worth it to use generator, but the question should be considered on a case by case basis. So the user should **consider** it and we could keep a `consider-using-generator` for those. What do you think ?

Reply:
@ngie-eign  I am using python 3.6.9 version.

I have tested @ngie-eign codes using `timeit`.
```
tuple(list(range(100)))
tuple(range(100))
set(list(range(100)))
set(range(100))

tuple(list(range(1000000)))
tuple(range(1000000))
set(list(range(1000000)))
set(range(1000000))
```

output:
```
1000000 loops, best of 3: 1.44 usec per loop
1000000 loops, best of 3: 0.927 usec per loop
100000 loops, best of 3: 2.72 usec per loop
1000000 loops, best of 3: 2.28 usec per loop


10 loops, best of 3: 45.9 msec per loop
10 loops, best of 3: 36.3 msec per loop
10 loops, best of 3: 93.9 msec per loop
10 loops, best of 3: 64.9 msec per loop
```

Clearly using the generator wins over the timing because the conversion is unnecessary. But the codes are not using list comprehension. I think we can create another case to check such unnecessary copies code.

Reply:
I also profiled the memory for `tuple` and `set` when using list comprehension.
Codes:
```
a = set([x % 7 for x in range(100)])
a = set(x % 7 for x in range(100))
a = set([x % 7 for x in range(1000000)])
a = set(x % 7 for x in range(1000000))

a = tuple([x % 7 for x in range(100)])
a = tuple(x % 7 for x in range(100))
a = tuple([x % 7 for x in range(1000000)])
a = tuple(x % 7 for x in range(1000000))
```
Results:
```
Maximum resident set size (kbytes): 9584
Maximum resident set size (kbytes): 9556
Maximum resident set size (kbytes): 17020
Maximum resident set size (kbytes): 9556

Maximum resident set size (kbytes): 9584
Maximum resident set size (kbytes): 9572
Maximum resident set size (kbytes): 24936
Maximum resident set size (kbytes): 17816
```

It is clear that a very long list/set will get the benefit from using generator :).

Reply:
@Pierre-Sassoulas : I think your suggestion about creating additional suggestions for using generators judiciously with other containers makes a lot of sense (seems like a good candidate for an I class message).

I also think removing the “consider” part in the checker makes a lot of sense too.

Reply:
```suggestion
        checked_call = ["any", "all", "set", "list", "tuple"]
```